### PR TITLE
Fix: Usage of ref maxoccurs and minoccurs attributes

### PR DESCRIPTION
**Pull request by ydaniju**
Consider this XSD:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<schema xmlns="http://www.w3.org/2001/XMLSchema"
  xmlns:EVT="http://ws.gematik.de/conn/EventService/v7.2"
  targetNamespace="http://ws.gematik.de/conn/EventService/v7.2" elementFormDefault="qualified"
  attributeFormDefault="unqualified" version="7.2.1">
  <simpleType name="SubscriptionIDType">
    <restriction base="string">
      <maxLength value="64" />
    </restriction>
  </simpleType>
  <element name="SubscriptionID" type="EVT:SubscriptionIDType"/>
  <element name="Subscribe">
    <complexType>
      <sequence>
        <element ref="EVT:Subscription" /> <!--Reference to Subscription without minOccurs or maxOccurs-->
      </sequence>
    </complexType>
  </element>
  <element name="GetSubscription">
    <complexType>
      <sequence>
        <element ref="EVT:SubscriptionID" minOccurs="0" />
      </sequence>
      <attribute name="mandant-wide" type="boolean" use="optional" default="false" />
    </complexType>
  </element>
  <element name="GetSubscriptionResponse">
    <complexType>
      <sequence>
        <element name="Subscriptions">
          <complexType>
            <sequence>
              <element ref="EVT:Subscription" minOccurs="0" maxOccurs="unbounded" /> <!--Reference to Subscription with minOccurs or maxOccurs-->
            </sequence>
          </complexType>
        </element>
      </sequence>
    </complexType>
  </element>
  <complexType name="SubscriptionType">
    <sequence>
      <element ref="EVT:SubscriptionID" minOccurs="0" />
      <element ref="EVT:TerminationTime" minOccurs="0" />
    </sequence>
  </complexType>
  <element name="Subscription" type="EVT:SubscriptionType" />
  <element name="TerminationTime" type="dateTime" />
</schema>
```

I expected `Subscriptions` to be returned as a List `Subscription`.

Sample response SOAP:

```xml
<?xml version='1.0' encoding='UTF-8'?>
<S:Envelope xmlns:S="http://schemas.xmlsoap.org/soap/envelope/">
    <S:Body>
        <ns11:GetSubscriptionResponse xmlns:ns2="http://ws.gematik.de/conn/ConnectorCommon/v5.0"
            xmlns:ns11="http://ws.gematik.de/conn/EventService/v7.2"
        >
            <ns2:Status>
                <ns2:Result>OK</ns2:Result>
            </ns2:Status>
            <ns11:Subscriptions>
                <ns11:Subscription>
                    <ns11:SubscriptionID>a24610ec-4069-4efc-a87e-5ab1d944ec5f</ns11:SubscriptionID>
                    <ns11:TerminationTime>2024-10-31T20:00:27.567+01:00</ns11:TerminationTime>
                </ns11:Subscription>
            </ns11:Subscriptions>
        </ns11:GetSubscriptionResponse>
    </S:Body>
</S:Envelope>

```

The current value being returned if there is only one subscription

```json
{
  "Status": {
    "Result": "OK"
  },
  "Subscriptions": {
    "Subscription": {
        "SubscriptionID": "a24610ec-4069-4efc-a87e-5ab1d944ec5f",
        "TerminationTime": "2024-10-31T20:00:27.567+01:00"
     }
  }
}

```

When ref is used to set maxoccurs, if the returned value is a single value, it will not be converted to an array.

Response I'm expecting.

```json
{
  "Status": {
    "Result": "OK"
  },
  "Subscriptions": {
    "Subscription": [
      {
        "SubscriptionID": "a24610ec-4069-4efc-a87e-5ab1d944ec5f",
        "TerminationTime": "2024-10-31T20:00:27.567+01:00"
      }
    ]
  }
}
```

### CODE REVIEW

**Comment by w666** [comment]
Hi @ydaniju, I assume this PR is not ready for review yet?

**Comment by ydaniju** [comment]
> Hi @ydaniju, I assume this PR is not ready for review yet?

Hey @w666 It's not ready yet. I have one failing test to fix right now

**Comment by ydaniju** [comment]
Now ready for review @w666

File: test/request-response-samples/GetSubscription__should_handle_ref_minOccurs_and_maxOccurs/response.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "Status": {
+    "Result": "OK"
+  },
+  "Subscriptions": {
```
Comment (w666) [inline test/request-response-samples/GetSubscription__should_handle_ref_minOccurs_and_maxOccurs/response.json:5]: Why `Subscriptions` is not an array, like it was described in [1168](https://github.com/vpulim/node-soap/pull/1168)
```json
  "Subscriptions": [{
    "Subscription": [
      {
        "SubscriptionID": "a24610ec-4069-4efc-a87e-5ab1d944ec5f",
        "TerminationTime": "2024-10-31T20:00:27.567+01:00"
      }
    ]
  }]
```

File: src/wsdl/elements.ts [hunk 2]
```diff
@@ -229,16 +229,16 @@ export class ElementElement extends Element {
...
       name += '[]';
     }
 
     if (xmlns && xmlns[TNS_PREFIX]) {
       this.$targetNamespace = xmlns[TNS_PREFIX];
     }
-    let type: any = this.$type || this.$ref;
-    if (type) {
-      type = splitQName(type);
+    const inferredType: string = this.$type || this.$ref;
```
Comment (w666) [inline src/wsdl/elements.ts:239]: What was the reason to rename this variable? Usually this is not a good practice to rename variables/function names without actual changes. 

Also, not sure that that type `string` is correct here? Is it possible that `$type` and `$ref` can be undefined or null?

I think this change is not related and should be reverted.

File: test/request-response-samples/GetSubscription__should_handle_ref_minOccurs_and_maxOccurs/response.json (hunk 1 above)
Comment (ydaniju) [reply test/request-response-samples/GetSubscription__should_handle_ref_minOccurs_and_maxOccurs/response.json:5]: - Making Subscriptions an array is not backward compatible. Without my change, we can still get an array of `Subscription`s if there are 2 or more subscriptions and it is returned as:

```xml
  "Subscriptions": {
    "Subscription": [
      {
        "SubscriptionID": "a24610ec-4069-4efc-a87e-5ab1d944ec5f",
        "TerminationTime": "2024-10-31T20:00:27.567+01:00"
      },
      {
        "SubscriptionID": "c24610ec-4069-4efc-a87e-5ab1d944ec5d",
        "TerminationTime": "2025-10-31T20:00:27.567+01:00"
      },
    ]
  }
```
- The XSD in #1168 is also quite different from the one here. The element with ref="tns:DummyList" suggests a list of a list of element with ref="DummyValues".
```xml
<!-- From #1168:  List of List of DummyValues-->
<xs:elements ref="tns:DummyList" minOccurs="1" maxOccurs="10" />
<xs:element name="DummyList">
  <xs:complexType> 
    <xs:sequence>
      <xs:element ref="DummyValues" minOccurs="1" maxOccurs="10" />
    </xs:sequence>
  </xs:complexType>
</xs:element>
```
The XSD in this PR suggests `Subscriptions` is a list of `Subscription`.

File: src/wsdl/elements.ts (hunk 2 above)
Comment (ydaniju) [reply src/wsdl/elements.ts:239]: > What was the reason to rename this variable? Usually this is not a good practice to rename variables/function names without actual changes.

I did this to better debug and can remove it. Main point is that type is assigned a string value and then is assigned to an object.

File: test/request-response-samples/GetSubscription__should_handle_ref_minOccurs_and_maxOccurs/response.json (hunk 1 above)
Comment (w666) [reply test/request-response-samples/GetSubscription__should_handle_ref_minOccurs_and_maxOccurs/response.json:5]: okay, that makes sense.

File: src/wsdl/elements.ts (hunk 2 above)
Comment (w666) [reply src/wsdl/elements.ts]: thanks

**Review by w666 — APPROVED** [review]

**Comment by w666** [comment]
@ydaniju is this fix urgent for you? 

There is another PR in progress that I would like to include into next release.
But if you need this ASAP, I can build a release earlier.

**Comment by ydaniju** [comment]
> @ydaniju is this fix urgent for you?
> 
> There is another PR in progress that I would like to include into next release. But if you need this ASAP, I can build a release earlier.

@w666 This is quite urgent. Would be great to have it sooner if it is not an inconvenience. Thank you.

**Comment by w666** [comment]
@ydaniju version 1.1.6 has been released
